external/xtorch_harmonics/xtorch_harmonics/xtorch_harmonics.py

```diff
@@ -0,0 +1,208 @@
+import warnings
+
+import numpy as np
+import torch
+import torch_harmonics
+import xarray as xr
+
+from typing import Hashable, Literal, TypeVar
+
+
+EQUIANGULAR_GRID = "equiangular"
+LEGENDRE_GAUSS_GRID = "legendre-gauss"
+LOBATTO_GRID = "lobatto"
+VALID_GRIDS = {
+    (EQUIANGULAR_GRID, EQUIANGULAR_GRID),
+    (LEGENDRE_GAUSS_GRID, LEGENDRE_GAUSS_GRID),
+    (LOBATTO_GRID, LOBATTO_GRID),
+    (EQUIANGULAR_GRID, LEGENDRE_GAUSS_GRID),
+    (LEGENDRE_GAUSS_GRID, EQUIANGULAR_GRID),
+}
+QUADRATURE_FUNCTIONS = {
+    EQUIANGULAR_GRID: torch_harmonics.quadrature.clenshaw_curtiss_weights,
+    LEGENDRE_GAUSS_GRID: torch_harmonics.quadrature.legendre_gauss_weights,
+    LOBATTO_GRID: torch_harmonics.quadrature.lobatto_weights,
+}
+LON_DIM = "grid_xt"
+LAT_DIM = "grid_yt"
+
+T_XarrayObject = TypeVar("T_XarrayObject", xr.DataArray, xr.Dataset)
+T_Grid = Literal[EQUIANGULAR_GRID, LEGENDRE_GAUSS_GRID, LOBATTO_GRID]
+
+
+def compute_quadrature_latitudes(n_lat: int, grid: T_Grid) -> np.ndarray:
+    if grid not in QUADRATURE_FUNCTIONS:
+        raise ValueError(f"Unrecognized grid {grid!r}.")
+
+    quadrature_function = QUADRATURE_FUNCTIONS[grid]
+    nodes, _ = quadrature_function(n_lat)
+    return np.rad2deg(np.arcsin(nodes))
+
+
+def compute_quadrature_longitudes(n_lon: int) -> np.ndarray:
+    bounds = np.linspace(0, 360, n_lon + 1, endpoint=True)
+    return (bounds[:-1] + bounds[1:]) / 2
+
+
+def _validate_quadrature_points(
+    obj: T_XarrayObject, forward_grid: T_Grid, lat_dim: Hashable, lon_dim: Hashable
+):
+    n_lat = obj.sizes[lat_dim]
+    n_lon = obj.sizes[lon_dim]
+
+    expected_latitudes = compute_quadrature_latitudes(n_lat, forward_grid)
```

Comment:
What if a dataset has latitudes in opposite order? I think it's still well posed to compute SHT, although for sure the sign of (some) coefficients will change.

Reply:
Good point, yes, this is another example where we might want more flexibility in validation.  I updated the check to allow this case too (in general this flexibility at least seems fair in the `roundtrip` case).